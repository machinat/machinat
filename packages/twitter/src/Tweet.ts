import TwitterUser from './User';
import polishMentionEntity from './utils/polishMentionEntity';
import polishUrlEntity from './utils/polishUrlEntity';
import polishMedia from './utils/polishMedia';
import type {
  RawTweet,
  Place,
  Coordinates,
  HashtagEntity,
  SymbolEntity,
  UrlEntity,
  MentionEntity,
  PollEntity,
  ReplyToTarget,
  Media,
  Photo,
  Video,
  AnimatedGif,
} from './types';

export default class Tweet {
  data: RawTweet;
  constructor(data: RawTweet) {
    this.data = data;
  }

  /** The id of the tweet */
  get id(): string {
    return this.data.id_str;
  }

  /** The user who create the tweet */
  get user(): TwitterUser {
    const rawUser = this.data.user;
    return new TwitterUser(rawUser.id_str, rawUser);
  }

  /** Time when this Tweet was created */
  get createdAt(): Date {
    return new Date(this.data.created_at);
  }

  /** The text of the status update. See [twitter-text](https://github.com/twitter/twitter-text/blob/master/rb/lib/twitter-text/regex.rb) for details on what characters are currently considered valid */
  get text(): string {
    return this.data.text;
  }

  /** Utility used to post the Tweet, as an HTML-formatted string. Tweets from the Twitter website have a source value of `'web'` */
  get source(): string {
    return this.data.source;
  }

  /** Indicates whether the value of the text parameter was truncated. In most cases, it's `false`  */
  get truncated(): boolean {
    return this.data.truncated;
  }

  /** The place that the tweet is associated (but not necessarily originating from) */
  get place(): null | Place {
    return this.data.place
      ? {
          id: this.data.place.id,
          url: this.data.place.url,
          type: this.data.place.place_type,
          name: this.data.place.name,
          fullName: this.data.place.full_name,
          countryCode: this.data.place.country_code,
          country: this.data.place.country,
          boundingCoordinates: this.data.place.bounding_box.coordinates.map(
            ([longitude, latitude]) => ({ longitude, latitude })
          ),
        }
      : null;
  }

  /** Represents the geographic location of this Tweet as reported by the user or client application */
  get coordinates(): null | Coordinates {
    if (!this.data.coordinates) {
      return null;
    }
    const [longitude, latitude] = this.data.coordinates.coordinates;
    return { longitude, latitude };
  }

  /** Indicates approximately how many times this Tweet has been quoted by Twitter users */
  get quoteCount(): undefined | number {
    return this.data.quote_count;
  }

  /** Number of times this Tweet has been replied to */
  get replyCount(): number {
    return this.data.reply_count;
  }

  /** Number of times this Tweet has been retweeted */
  get retweetCount(): number {
    return this.data.retweet_count;
  }

  /** Indicates approximately how many times this Tweet has been liked by Twitter users */
  get favoriteCount(): undefined | number {
    return this.data.favorite_count;
  }

  /** Indicates whether this Tweet has been liked by the authenticating user */
  get favorited(): boolean {
    return this.data.favorited;
  }

  /** Indicates whether this Tweet has been Retweeted by the authenticating user */
  get retweeted(): boolean {
    return this.data.retweeted;
  }

  /** This field only surfaces when a Tweet contains a link. The meaning of the field doesnâ€™t pertain to the Tweet content itself, but instead it is an indicator that the URL contained in the Tweet may contain content or media identified as sensitive content */
  get possiblySensitive(): undefined | boolean {
    return this.data.possibly_sensitive;
  }

  /** Indicates the maximum value of the filter_level parameter which may be used and still stream this Tweet. So a value of medium will be streamed on none, low, and medium streams */
  get filterLevel(): string {
    return this.data.filter_level;
  }

  /** When present, indicates a BCP 47 language identifier corresponding to the machine-detected language of the Tweet text, or `'und'` if no language could be detected */
  get languageCode(): undefined | string {
    return this.data.lang;
  }

  /** Represents hashtags which have been parsed out of the Tweet text */
  get hashtags(): HashtagEntity[] {
    return this.data.entities.hashtags;
  }

  get symbols(): SymbolEntity[] {
    return this.data.entities.symbols;
  }

  /** Represents URLs included in the text of a Tweet */
  get urls(): UrlEntity[] {
    return this.data.entities.urls.map(polishUrlEntity);
  }

  get userMentions(): MentionEntity[] {
    return this.data.entities.user_mentions.map(polishMentionEntity);
  }

  /** Represents Twitter Polls included in the Tweet */
  get poll(): null | PollEntity {
    if (!this.data.entities.polls || this.data.entities.polls.length === 0) {
      return null;
    }
    const raw = this.data.entities.polls[0];
    return {
      options: raw.options,
      endTime: new Date(raw.end_datetime),
      durationMinutes: raw.duration_minutes,
    };
  }

  get replyTo(): null | ReplyToTarget {
    const tweetId = this.data.in_reply_to_status_id_str;
    if (!tweetId) {
      return null;
    }

    const userId = this.data.in_reply_to_user_id_str as string;
    const screenName = this.data.in_reply_to_screen_name as string;
    return {
      tweetId,
      userId,
      screenName,
      user: new TwitterUser(userId),
    };
  }

  /** The Tweet object of the original Tweet that was quoted */
  get quotedStatus(): null | Tweet {
    return this.data.quoted_status ? new Tweet(this.data.quoted_status) : null;
  }

  /** The original Tweet that was retweeted. Note that retweets of retweets do not show representations of the intermediary retweet, but only the original Tweet */
  get retweetedStatus(): null | Tweet {
    return this.data.retweeted_status
      ? new Tweet(this.data.retweeted_status)
      : null;
  }

  get video(): null | Video {
    if (this.data.extended_entities.media[0]?.type !== 'video') {
      return null;
    }
    return polishMedia(this.data.extended_entities.media[0]) as Video;
  }

  get animatedGif(): null | AnimatedGif {
    if (this.data.extended_entities.media[0]?.type !== 'animated_gif') {
      return null;
    }
    return polishMedia(this.data.extended_entities.media[0]) as AnimatedGif;
  }

  get photos(): null | Photo[] {
    if (this.data.extended_entities.media[0]?.type !== 'photo') {
      return null;
    }
    return this.data.extended_entities.media.map(polishMedia) as Photo[];
  }

  get media(): null | Media[] {
    return this.data.extended_entities.media.map(polishMedia);
  }
}
