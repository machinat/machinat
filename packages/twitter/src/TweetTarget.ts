import type { SociablyThread, UniqueOmniIdentifier } from '@sociably/core';
import type { MarshallableInstance } from '@sociably/core/base/Marshaler';
import TwitterUser from './User';
import { TWITTER, TWTR } from './constant';

type SerializedTweetTarget = {
  agent: string;
  tweet?: string;
};

const DEFAULT_FEED_SIGN = '-';

export default class TwitterTweetTarget
  implements SociablyThread, MarshallableInstance<SerializedTweetTarget>
{
  static typeName = 'TwtrTweetTarget';
  static fromJSONValue(val: SerializedTweetTarget): TwitterTweetTarget {
    return new TwitterTweetTarget(val.agent, val.tweet);
  }

  platform = TWITTER;
  /** The id of the agent user */
  agentId: string;
  /** The tweet to reply. If it's empty, the tweet is created under agent user's page  */
  tweetId?: string;

  constructor(agentId: string, tweetId?: string) {
    this.agentId = agentId;
    this.tweetId = tweetId;
  }

  get uniqueIdentifier(): UniqueOmniIdentifier {
    return {
      platform: TWITTER,
      scopeId: this.agentId,
      id: this.tweetId || DEFAULT_FEED_SIGN,
    };
  }

  get uid(): string {
    return `${TWTR}.${this.agentId}.${this.tweetId || DEFAULT_FEED_SIGN}`;
  }

  get agent(): TwitterUser {
    return new TwitterUser(this.agentId);
  }

  // eslint-disable-next-line class-methods-use-this
  typeName(): string {
    return TwitterTweetTarget.typeName;
  }

  toJSONValue(): SerializedTweetTarget {
    return { agent: this.agentId, tweet: this.tweetId };
  }
}
