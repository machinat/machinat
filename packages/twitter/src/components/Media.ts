import { NativeElement } from '@sociably/core';
import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import makeTwitterComponent from '../utils/makeTwitterComponent';
import { TwitterSegmentValue, TwitterComponent, MediaType } from '../types';

/**
 * @category Props
 */
export type MediaProps = {
  /** Set to true if media asset will be reused for multiple Direct Messages. Default is false */
  shared?: boolean;
  /** The tag to store media asset with */
  assetTag?: string;
  /** A string enum value which identifies a media usecase. This identifier is used to enforce usecase specific constraints (e.g. file size, video duration) and enable advanced features */
  mediaCategory?: string;
  /** A comma-separated list of user IDs to set as additional owners allowed to use the returned media_id in Tweets or Cards. Up to 100 additional owners may be specified */
  additionalOwners?: string[];
  /** The media file url */
  url?: string;
  /** The uploaded media id */
  mediaId?: string;
  /** The file content to be upload */
  fileData?: Buffer | NodeJS.ReadableStream;
  /** The content type of the file. Required when uploading using `file` */
  fileType?: string;
  /** The file size in bytes. Required when uploading using `file` */
  fileSize?: number;
};

const makeMediaComponent = (name: string, mediaType: MediaType) =>
  makeTwitterComponent(
    {
      [name](
        node: NativeElement<MediaProps, TwitterComponent<MediaProps>>,
        path: string
      ): UnitSegment<TwitterSegmentValue>[] {
        const {
          shared,
          assetTag,
          url,
          mediaId,
          fileData,
          fileSize,
          fileType,
          mediaCategory,
          additionalOwners,
        } = node.props;

        if (
          !(url || mediaId || fileData) ||
          (url && mediaId) ||
          (url && fileData) ||
          (fileData && mediaId)
        ) {
          throw new TypeError(
            'there should be exactly one of "url", "mediaId" or "fileData" prop'
          );
        }
        if (fileData && (!fileType || !fileSize)) {
          throw new TypeError(
            '"fileType" and "fileSize" props are required when using "fileData" source'
          );
        }

        const params = {
          shared: shared ? 'true' : undefined,
          media_category: mediaCategory,
          additional_owners: additionalOwners
            ? additionalOwners.join(',')
            : undefined,
        };

        return [
          makeUnitSegment(node, path, {
            type: 'media',
            attachment: mediaId
              ? {
                  type: mediaType,
                  source: { type: 'id', id: mediaId, params },
                }
              : url
              ? {
                  type: mediaType,
                  source: { type: 'url', url, params },
                }
              : fileData
              ? {
                  type: mediaType,
                  source: {
                    type: 'file',
                    params: {
                      ...params,
                      total_bytes: fileSize,
                      media_type: fileType,
                    },
                    fileData,
                    assetTag,
                  },
                }
              : (null as never),
          }),
        ];
      },
    }[name]
  );

/**
 * Send a photo by direct message or attach it to a tweet
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).
 */
export const Photo: TwitterComponent<
  MediaProps,
  UnitSegment<TwitterSegmentValue>
> = makeMediaComponent('Photo', 'photo');

/**
 * Send a video by direct message or attach it to a tweet
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).
 */
export const Video: TwitterComponent<
  MediaProps,
  UnitSegment<TwitterSegmentValue>
> = makeMediaComponent('Video', 'video');

/**
 * Send an animated gif by direct message or attach it to a tweet
 * @category Component
 * @props {@link MediaProps}
 * @guides Check official [guide](https://developer.twitter.com/en/docs/twitter-api/v1/media/upload-media/overview).
 */
export const AnimatedGif: TwitterComponent<
  MediaProps,
  UnitSegment<TwitterSegmentValue>
> = makeMediaComponent('AnimatedGif', 'animated_gif');
