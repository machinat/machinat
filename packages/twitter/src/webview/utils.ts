import type { ContextDetails } from '@sociably/auth';
import TwitterChat from '../Chat';
import TwitterUser from '../User';
import UserProfile from '../UserProfile';
import type { TwitterAuthContext, TwitterAuthData } from './types';

// eslint-disable-next-line import/prefer-default-export
export const getAuthContextDetails = ({
  agent: agentId,
  user: { id: userId, data: rawUserData },
}: TwitterAuthData): ContextDetails<TwitterAuthContext> => {
  return {
    agentId,
    channel: new TwitterUser(agentId),
    user: new TwitterUser(userId, rawUserData),
    thread: new TwitterChat(agentId, userId),
    userProfile: new UserProfile(rawUserData),
  };
};
