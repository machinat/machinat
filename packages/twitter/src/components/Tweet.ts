import type { SociablyNode } from '@sociably/core';
import { formatNode } from '@sociably/core/utils';
import { makeUnitSegment, UnitSegment } from '@sociably/core/renderer';
import makeTwitterComponent from '../utils/makeTwitterComponent.js';
import createTweetSegmentValue from '../utils/createTweetSegmentValue.js';
import splitTweetText from '../utils/splitTweetText.js';
import {
  TwitterSegmentValue,
  TwitterComponent,
  MediaAttachment,
} from '../types.js';

/** @category Props */
export type TweetProps = {
  /** Texual content of the tweet */
  children?: SociablyNode;
  /**
   * Tweets a link directly to a Direct Message conversation with an account. By
   * default it links to the agent who create the tweet. You can directly pass
   * the URL string or an object specifying the URL queries.
   */
  directMessageLink?:
    | boolean
    | string
    | {
        text?: string;
        recipientId?: string;
        welcomeMessageId?: string;
      };
  /** Place ID being attached to the Tweet for geo location */
  placeId?: string;
  /**
   * Link to the Tweet being quoted. This is mutually exclusive from `media` and
   * `poll`
   */
  quoteTweetId?: string;
  /**
   * Attach media to the tweet. Should contain only {@link Media}. This is
   * mutually exclusive from `poll` and `quoteTweetId`
   */
  media?: SociablyNode;
  /**
   * A list of User IDs being tagged in the Tweet with Media. If the user you're
   * tagging doesn't have photo-tagging enabled, their names won't show up in
   * the list of tagged users even though the Tweet is successfully created
   */
  tagUsersInMedia?: string[];
  /**
   * Attach a poll to the tweet. This is mutually exclusive from `media` and
   * `quoteTweetId`
   */
  poll?: {
    /** Duration of the poll in minutes for a Tweet with a poll */
    durationMinutes: number;
    /** A list of poll options for a Tweet with a poll */
    options: string[];
  };
  /**
   * A list of User IDs to be excluded from the reply Tweet thus removing a user
   * from a thread
   */
  excludeUsersInReply?: string[];
  /** Tweet exclusively for Super Followers. */
  superFollowersOnly?: boolean;
  /** Indicate who can reply to the Tweet. Default to `'everyone'` */
  replySetting?: 'everyone' | 'following' | 'mentionedUsers';
};

/**
 * Send direct messages with metadata
 *
 * @category Component
 * @props {@link TweetProps}
 * @guides Check official [guides](https://developer.twitter.com/en/docs/twitter-api/v1/direct-messages/api-features).
 */
export const Tweet: TwitterComponent<
  TweetProps,
  UnitSegment<TwitterSegmentValue>
> = makeTwitterComponent(async function Tweet(node, path, render) {
  const {
    children,
    directMessageLink,
    placeId,
    media,
    tagUsersInMedia,
    poll,
    quoteTweetId,
    excludeUsersInReply,
    superFollowersOnly,
    replySetting,
  } = node.props;

  if ((media && poll) || (media && quoteTweetId) || (poll && quoteTweetId)) {
    throw new TypeError(
      'there should be exactly one of "media", "poll" or "quoteTweetId" prop',
    );
  }

  const [contentSegments, mediaSegments] = await Promise.all([
    render<TwitterSegmentValue>(children, '.children'),
    render<TwitterSegmentValue>(media, '.media'),
  ]);

  for (const segment of contentSegments || []) {
    if (segment.type !== 'text') {
      throw new TypeError(
        `non-texual element ${formatNode(
          segment.node,
        )} can't be placed under <Tweet/>`,
      );
    }
  }

  let mediaAttachments: null | MediaAttachment[] = null;
  if (mediaSegments) {
    mediaAttachments = [];

    for (const segment of mediaSegments) {
      if (
        (segment.type !== 'unit' && segment.type !== 'raw') ||
        segment.value.type !== 'media'
      ) {
        throw new TypeError(
          `${formatNode(segment.node)} can't be placed in "media" prop`,
        );
      }

      mediaAttachments.push(segment.value.attachment);
    }
  }

  if (!contentSegments && !mediaAttachments) {
    throw new TypeError(`no text or media in <Tweet/>`);
  }

  const content = contentSegments?.[0].value as string;
  const splitedContent = splitTweetText(content);

  const segments = [
    makeUnitSegment(
      node,
      path,
      createTweetSegmentValue(
        {
          text: splitedContent?.[0],
          reply: excludeUsersInReply
            ? {
                in_reply_to_tweet_id: '',
                exclude_reply_user_ids: excludeUsersInReply,
              }
            : undefined,
          media: tagUsersInMedia
            ? {
                media_ids: [],
                tagged_user_ids: tagUsersInMedia,
              }
            : undefined,
          quote_tweet_id: quoteTweetId,
          poll: poll
            ? {
                options: poll.options,
                duration_minutes: poll.durationMinutes,
              }
            : undefined,
          geo: placeId ? { place_id: placeId } : undefined,
          reply_settings: replySetting,
          for_super_followers_only: superFollowersOnly,
          direct_message_deep_link: undefined,
        },
        mediaAttachments,
        directMessageLink,
      ),
    ),
  ];

  if (splitedContent) {
    segments.push(
      ...splitedContent
        .slice(1)
        .map((text) =>
          makeUnitSegment(node, path, createTweetSegmentValue({ text })),
        ),
    );
  }

  return segments;
});
