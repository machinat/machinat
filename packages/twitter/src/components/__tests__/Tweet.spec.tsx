import Machinat from '@machinat/core';
import { isNativeType } from '@machinat/core/utils';
import Renderer from '@machinat/core/renderer';
import TweetTarget from '../../TweetTarget';
import { Tweet } from '../Tweet';
import { Photo } from '../Media';

const renderer = new Renderer('twitter', async () => null);
const render = (element) => renderer.render(element, null, null);

it('is a valid Component', () => {
  expect(typeof Tweet).toBe('function');
  expect(isNativeType(<Tweet />)).toBe(true);
  expect(Tweet.$$platform).toBe('twitter');
});

test("rendering to agent's page", async () => {
  const segments = await render(
    <Tweet quoteTweetId="67890" replySetting="following" superFollowersOnly>
      Hello World
    </Tweet>
  );
  expect(segments).toMatchSnapshot();
  const { request, accomplishRequest } = (segments as any)[0].value;
  expect(accomplishRequest(new TweetTarget('12345'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "2/tweets",
      "method": "POST",
      "parameters": Object {
        "direct_message_deep_link": undefined,
        "for_super_followers_only": true,
        "geo": undefined,
        "media": undefined,
        "poll": undefined,
        "quote_tweet_id": "67890",
        "reply": undefined,
        "reply_settings": "following",
        "text": "Hello World",
      },
    }
  `);
});

test('rendering as a reply', async () => {
  const segments = await render(
    <Tweet
      placeId="55555"
      poll={{ durationMinutes: 60, options: ['foo', 'bar', 'baz'] }}
      excludeUsersInReply={['56789']}
    >
      Hello World
    </Tweet>
  );
  expect(segments).toMatchSnapshot();
  const { request, accomplishRequest } = (segments as any)[0].value;
  expect(accomplishRequest(new TweetTarget('12345', '67890'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "2/tweets",
      "method": "POST",
      "parameters": Object {
        "direct_message_deep_link": undefined,
        "for_super_followers_only": undefined,
        "geo": Object {
          "place_id": "55555",
        },
        "media": undefined,
        "poll": Object {
          "duration_minutes": 60,
          "options": Array [
            "foo",
            "bar",
            "baz",
          ],
        },
        "quote_tweet_id": undefined,
        "reply": Object {
          "exclude_reply_user_ids": Array [
            "56789",
          ],
          "in_reply_to_tweet_id": "67890",
        },
        "reply_settings": undefined,
        "text": "Hello World",
      },
    }
  `);
});

test('rendering with media', async () => {
  const segments = await render(
    <Tweet
      superFollowersOnly
      tagUsersInMedia={['56789']}
      media={
        <>
          <Photo mediaId="123" />
          <Photo url="http://foo.bar/baz.png" />
          <Photo
            file={Buffer.from('foo')}
            fileSize={123}
            fileType="image/jpeg"
          />
        </>
      }
    />
  );
  expect(segments).toMatchSnapshot();
  const { request, accomplishRequest } = (segments as any)[0].value;
  expect(
    accomplishRequest(new TweetTarget('12345'), request, [
      '12345',
      '23456',
      '34567',
    ])
  ).toMatchInlineSnapshot(`
    Object {
      "href": "2/tweets",
      "method": "POST",
      "parameters": Object {
        "direct_message_deep_link": undefined,
        "for_super_followers_only": true,
        "geo": undefined,
        "media": Object {
          "media_ids": Array [
            "12345",
            "23456",
            "34567",
          ],
          "tagged_user_ids": Array [
            "56789",
          ],
        },
        "poll": undefined,
        "quote_tweet_id": undefined,
        "reply": undefined,
        "reply_settings": undefined,
        "text": undefined,
      },
    }
  `);
});

test("rendering to agent's page", async () => {
  let segments = await render(<Tweet directMessageLink>Hello World</Tweet>);
  let { request, accomplishRequest } = (segments as any)[0].value;
  expect(accomplishRequest(new TweetTarget('12345'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "2/tweets",
      "method": "POST",
      "parameters": Object {
        "direct_message_deep_link": "https://twitter.com/messages/compose?recipient_id=12345",
        "for_super_followers_only": undefined,
        "geo": undefined,
        "media": undefined,
        "poll": undefined,
        "quote_tweet_id": undefined,
        "reply": undefined,
        "reply_settings": undefined,
        "text": "Hello World",
      },
    }
  `);

  segments = await render(
    <Tweet directMessageLink="https://twitter.com/messages/compose?recipient_id=23456&text=boo">
      Hello World
    </Tweet>
  );
  ({ request, accomplishRequest } = (segments as any)[0].value);
  expect(accomplishRequest(new TweetTarget('12345'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "2/tweets",
      "method": "POST",
      "parameters": Object {
        "direct_message_deep_link": "https://twitter.com/messages/compose?recipient_id=23456&text=boo",
        "for_super_followers_only": undefined,
        "geo": undefined,
        "media": undefined,
        "poll": undefined,
        "quote_tweet_id": undefined,
        "reply": undefined,
        "reply_settings": undefined,
        "text": "Hello World",
      },
    }
  `);

  segments = await render(
    <Tweet
      directMessageLink={{ text: 'hello world', welcomeMessageId: '22222' }}
    >
      Hello World
    </Tweet>
  );
  ({ request, accomplishRequest } = (segments as any)[0].value);
  expect(accomplishRequest(new TweetTarget('12345'), request, null))
    .toMatchInlineSnapshot(`
    Object {
      "href": "2/tweets",
      "method": "POST",
      "parameters": Object {
        "direct_message_deep_link": "https://twitter.com/messages/compose?recipient_id=12345&welcome_message_id=22222&text=hello%20world",
        "for_super_followers_only": undefined,
        "geo": undefined,
        "media": undefined,
        "poll": undefined,
        "quote_tweet_id": undefined,
        "reply": undefined,
        "reply_settings": undefined,
        "text": "Hello World",
      },
    }
  `);
});

it('throw if multiple attchments received', async () => {
  await expect(
    render(
      <Tweet
        quoteTweetId="23456"
        poll={{ durationMinutes: 1, options: ['2', '3'] }}
      >
        Hello
      </Tweet>
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"there should be exactly one of \\"media\\", \\"poll\\" or \\"quoteTweetId\\" prop"`
  );

  await expect(
    render(<Tweet media={<Photo mediaId="33333" />} quoteTweetId="23456" />)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"there should be exactly one of \\"media\\", \\"poll\\" or \\"quoteTweetId\\" prop"`
  );
});

it('throw if non-textual content in children', async () => {
  await expect(
    render(
      <Tweet>
        <Photo mediaId="44444" />
      </Tweet>
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-texual element <Photo /> can't be placed under <Tweet/>"`
  );

  await expect(
    render(
      <Tweet>
        <Machinat.Pause />
      </Tweet>
    )
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"non-texual element <Pause /> can't be placed under <Tweet/>"`
  );
});

it('throw if invalid media received', async () => {
  await expect(
    render(<Tweet media={'foo' as never} />)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"\\"foo\\" can't be placed in \\"media\\" prop"`
  );

  await expect(
    render(<Tweet media={<Machinat.Pause />} />)
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `"<Pause /> can't be placed in \\"media\\" prop"`
  );
});

it('throw if no content available', async () => {
  await expect(render(<Tweet />)).rejects.toThrowErrorMatchingInlineSnapshot(
    `"no text or media in <Tweet/>"`
  );
});
