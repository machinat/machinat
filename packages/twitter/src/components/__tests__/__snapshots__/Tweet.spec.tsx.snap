// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering as a reply 1`] = `
Array [
  Object {
    "node": <Tweet
      excludeUsersInReply={
        Array [
          "56789",
        ]
      }
      placeId="55555"
      poll={
        Object {
          "durationMinutes": 60,
          "options": Array [
            "foo",
            "bar",
            "baz",
          ],
        }
      }
    >
      Hello World
    </Tweet>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "2/tweets",
        "method": "POST",
        "parameters": Object {
          "direct_message_deep_link": undefined,
          "for_super_followers_only": undefined,
          "geo": Object {
            "place_id": "55555",
          },
          "media": undefined,
          "poll": Object {
            "duration_minutes": 60,
            "options": Array [
              "foo",
              "bar",
              "baz",
            ],
          },
          "quote_tweet_id": undefined,
          "reply": Object {
            "exclude_reply_user_ids": Array [
              "56789",
            ],
            "in_reply_to_tweet_id": "",
          },
          "reply_settings": undefined,
          "text": "Hello World",
        },
      },
      "type": "tweet",
    },
  },
]
`;

exports[`rendering to agent's page 1`] = `
Array [
  Object {
    "node": <Tweet
      quoteTweetId="67890"
      replySetting="following"
      superFollowersOnly={true}
    >
      Hello World
    </Tweet>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "2/tweets",
        "method": "POST",
        "parameters": Object {
          "direct_message_deep_link": undefined,
          "for_super_followers_only": true,
          "geo": undefined,
          "media": undefined,
          "poll": undefined,
          "quote_tweet_id": "67890",
          "reply": undefined,
          "reply_settings": "following",
          "text": "Hello World",
        },
      },
      "type": "tweet",
    },
  },
]
`;

exports[`rendering with media 1`] = `
Array [
  Object {
    "node": <Tweet
      media={
        <Machinat.Fragment>
          <Photo
            mediaId="123"
          />
          <Photo
            url="http://foo.bar/baz.png"
          />
          <Photo
            fileData={
              Object {
                "data": Array [
                  102,
                  111,
                  111,
                ],
                "type": "Buffer",
              }
            }
            fileSize={123}
            fileType="image/jpeg"
          />
        </Machinat.Fragment>
      }
      superFollowersOnly={true}
      tagUsersInMedia={
        Array [
          "56789",
        ]
      }
    />,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": Array [
        Object {
          "id": "123",
          "parameters": Object {
            "additional_owners": undefined,
            "media_category": undefined,
            "shared": undefined,
          },
          "type": "id",
        },
        Object {
          "parameters": Object {
            "additional_owners": undefined,
            "media_category": "tweet_image",
            "shared": undefined,
          },
          "type": "url",
          "url": "http://foo.bar/baz.png",
        },
        Object {
          "assetTag": undefined,
          "fileData": Object {
            "data": Array [
              102,
              111,
              111,
            ],
            "type": "Buffer",
          },
          "parameters": Object {
            "additional_owners": undefined,
            "media_category": "tweet_image",
            "media_type": "image/jpeg",
            "shared": undefined,
            "total_bytes": 123,
          },
          "type": "file",
        },
      ],
      "request": Object {
        "href": "2/tweets",
        "method": "POST",
        "parameters": Object {
          "direct_message_deep_link": undefined,
          "for_super_followers_only": true,
          "geo": undefined,
          "media": Object {
            "media_ids": Array [],
            "tagged_user_ids": Array [
              "56789",
            ],
          },
          "poll": undefined,
          "quote_tweet_id": undefined,
          "reply": undefined,
          "reply_settings": undefined,
          "text": undefined,
        },
      },
      "type": "tweet",
    },
  },
]
`;

exports[`spliting long content into tweet thread 1`] = `
Array [
  Object {
    "node": <Tweet
      placeId="98765"
      quoteTweetId="67890"
      replySetting="following"
      superFollowersOnly={true}
    >
      國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國囧國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國
    </Tweet>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "2/tweets",
        "method": "POST",
        "parameters": Object {
          "direct_message_deep_link": undefined,
          "for_super_followers_only": true,
          "geo": Object {
            "place_id": "98765",
          },
          "media": undefined,
          "poll": undefined,
          "quote_tweet_id": "67890",
          "reply": undefined,
          "reply_settings": "following",
          "text": "國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國",
        },
      },
      "type": "tweet",
    },
  },
  Object {
    "node": <Tweet
      placeId="98765"
      quoteTweetId="67890"
      replySetting="following"
      superFollowersOnly={true}
    >
      國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國囧國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國
    </Tweet>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "2/tweets",
        "method": "POST",
        "parameters": Object {
          "text": "國國國國國國國國國國國國國國國國囧國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國",
        },
      },
      "type": "tweet",
    },
  },
  Object {
    "node": <Tweet
      placeId="98765"
      quoteTweetId="67890"
      replySetting="following"
      superFollowersOnly={true}
    >
      國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國囧國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國 國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國
    </Tweet>,
    "path": "$",
    "type": "unit",
    "value": Object {
      "accomplishRequest": [Function],
      "mediaSources": null,
      "request": Object {
        "href": "2/tweets",
        "method": "POST",
        "parameters": Object {
          "text": "國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國國",
        },
      },
      "type": "tweet",
    },
  },
]
`;
